source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,179 @@
+FilterHeadersStatus GzipFilter::decodeHeaders(HeaderMap& headers, bool) {
+  // TODO(gsagula): The current implementation checks for the presence of 'gzip' and if the same
+  // is followed by Qvalue. Since gzip is the only available content-encoding in Envoy at the
+  // moment, order/priority of preferred server encodings is disregarded (RFC2616-14.3).
+  skip_compression_ =
+      !(headers.AcceptEncoding() &&
+        std::regex_search(headers.AcceptEncoding()->value().c_str(), acceptEncodingRegex()));
+  return FilterHeadersStatus::Continue;
+}
+
+FilterHeadersStatus GzipFilter::encodeHeaders(HeaderMap& headers, bool end_stream) {
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isContentTypeAllowed(headers) && isCacheControlAllowed(headers) && isEtagAllowed(headers) &&
+      isLastModifiedAllowed(headers) && isTransferEncodingAllowed(headers) &&
+      !headers.ContentEncoding()) {
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(Http::Headers::get().ContentEncodingValues.Gzip);
+    compressor_.init(config_->compressionLevel(), config_->compressionStrategy(),
+                     config_->windowBits(), config_->memoryLevel());
+  } else {
+    skip_compression_ = true;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_compression_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t n_data{data.length()};
+
+  if (n_data) {
+    compressor_.compress(data, compressed_data_);
+  }
+
+  if (end_stream) {
+    compressor_.flush(compressed_data_);
+  }
+
+  if (compressed_data_.length()) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::isMinimumContentLength(HeaderMap& headers) const {
+  if (headers.ContentLength()) {
+    uint64_t content_length{};
+    StringUtil::atoul(headers.ContentLength()->value().c_str(), content_length);
+    return content_length >= config_->minimumLength();
+  }
+
+  return (headers.TransferEncoding() &&
+          headers.TransferEncoding()->value().find(
+              Http::Headers::get().TransferEncodingValues.Chunked.c_str()));
+}
+
+bool GzipFilter::isContentTypeAllowed(HeaderMap& headers) const {
+  if (headers.ContentType() && !config_->contentTypeValues().empty()) {
+    std::string content_type{headers.ContentType()->value().c_str()};
+    StringUtil::eraseFrom(content_type, ";");
+    StringUtil::trim(content_type);
+    return config_->contentTypeValues().find(content_type) != config_->contentTypeValues().end();
+  }
+
+  return true;
+}
+
+bool GzipFilter::isCacheControlAllowed(HeaderMap& headers) const {
+  if (headers.CacheControl() && !config_->cacheControlValues().empty()) {
+    auto header_values = StringUtil::split(headers.CacheControl()->value().c_str(), ',');
```

Comment:
don't forget to trim the tokens.  If you just split on "," you'll wind up with tokens like " max-age=0" and " public".  For some reason, it's common to have "public, max-age=300" in headers.

I'm not really sure if splitting off the max-age value is really a concern.  E.g. would someone want to say "I'm willing to gzip, but only if it's cached less than 5 minutes"?

But you need to strip the whitespace so you can find "no-transform".